Generalize trace validation of byte_extract_exprt to include all endiannesses in JBMC

jeannielynnmoulton opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add can_cast_expr for byte_extract_exprt

This is needed so that we can use expr_try_dynamic_cast with
byte_extract_exprts

Commit:
Generalize checks of byte_extract to include all endiannesses

It is a little bit tidier to use can_cast_expr and expr_try_dynamic_cast
instead of checking the id and also checks for both endiannesses, making
the trace validation more robust.

Commit:
Unit tests for casting byte_extract_exprt

Adds tests for both can_cast_expr and expr_try_dynamic_cast

Commit:
Update unit tests for trace validation in JBMC

These unit tests now include byte extract exprts with big endianness

owen-mc-diffblue commented on unit/util/expr_cast/expr_cast.cpp:
```diff
@@ -141,6 +142,67 @@ SCENARIO("expr_dynamic_cast",
+  }
+  GIVEN("A byte extract expression with big endianness")
+  {
+    auto byte = byte_extract_exprt(ID_byte_extract_little_endian);
```
You mean `big_endian`

thomasspriggs commented on src/util/byte_operators.h:
```diff
@@ -57,6 +57,13 @@ class byte_extract_exprt:public binary_exprt
+  return base.id() == ID_byte_extract_little_endian ||
+         base.id() == ID_byte_extract_big_endian;
+}
+
```
:pick: I guess we should probably implement `validate_expr` at the same time as `can_cast_expr`, for consistency with other classes which inherit from `binary_exprt`.
```
inline void validate_expr(const byte_extract_exprt &value)
{
  validate_operands(value, 2, "byte extract must have two operands");
}
```

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b46d3b775122222fa742c03e4c5208203db2ffbb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117287261

Commit:
Add override of validate_expr

Because we have can_can_expr for byte_extract, we should have an
override for validate_expr

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 86b4fb3f2fd39afae9aac091fb828cc000398d58).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117298045

smowton commented on jbmc/src/java_bytecode/java_trace_validation.cpp:
```diff
@@ -251,23 +251,23 @@ static void check_rhs_assumptions(
   }
   // check byte extract rhs structure
-  else if(rhs.id() == ID_byte_extract_little_endian)
+  else if(const auto byte = expr_try_dynamic_cast<byte_extract_exprt>(rhs))
```
byte_extract_expr

smowton reviewed (approved):
One nitpick otherwise lgtm
